enhancements/installer/bootstrapping-clusters-with-capi-providers.md

```diff
@@ -0,0 +1,646 @@
+      - controlplane
+      - node
+      toPort: 10259
+    cni:
+      cniIngressRules:
+      - description: ICMP
+        fromPort: -1
+        protocol: icmp
+        toPort: -1
+       # several rules omitted for brevity
+      - description: Service node ports (UDP)
+        fromPort: 30000
+        protocol: udp
+        toPort: 32767
+    vpc:
+      availabilityZoneSelection: Ordered
+      availabilityZoneUsageLimit: 3
+      cidrBlock: 10.0.0.0/16
+  region: us-east-2
+  s3Bucket:
+    name: openshift-bootstrap-data-capi-example-kbdpg
+    presignedURLDuration: 1h0m0s
+status:
+  networkStatus:
+    apiServerElb:
+      attributes: {}
+  ready: false
+```
+
+This manifest describes the network, load balancers, security group rules, etc. that will be used or provisioned for the target cluster.
+Control plane machines are provisioned by [AWSMachines][AWSMachines] & [CAPI Machines][CAPIMachines] manifests.
+
+##### Additional Infrastructure
+
+The Installer generates any additional infrastructure that is needed but not handled by the CAPI provider, either because it is out of scope (e.g. IAM, DNS), or not adopted upstream (e.g. split-horizon load balancers).
+The Installer codebase provides hooks into the provisioning lifecycle that can be used to provision resources using direct SDK calls or other tooling.
+
+In most cases, teams are encouraged in discussing and building issues in the respective upstream repositories first, e.g. `cluster-api` or `cluster-api-provider-aws` and only
+use the following hook as a fallback path.
+
+Provisioning of resources with CAPI is divided into two stages: _infrastructure_ (all non-machine related manifests) and _control plane_,
+with hooks before, in between, and after:
+
+* PreProvision Hook - handle prequisites: e.g. IAM role creation, OS Disk upload
+* Provision Infrastructure - create `cluster` manifest on local control plane
+* InfraReady Hook - handle requirements in between cluster infrastructure and machine provisioning
+* Ignition Hook - similar to InfraReady but specifically for generating (bootstrap) ignition
+* Provision Machines - create bootstrap ignition and control-plane machines
+* PostProvision Hook - post-provision requirements, such as DNS record creation
+
+##### Bootstrap Resources
+
+OpenShift will continue to use the same bootstrapping process, as described in the Installer
+[Cluster Installation Process][install-overview].
+The Bootstrap Machine is created like the other control plane nodes, by defining a CAPI Machine resource.
+Accordingly, the bootstrap machine can be deleted by simply deleting the object. This works for all platforms:
+
+```go
+	if err := c.Delete(ctx, &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      capiutils.GenerateBoostrapMachineName(metadata.InfraID),
+			Namespace: capiutils.Namespace,
+		},
+	});
+```
+
+Further work needs to be done to determine how to delete additional bootstrap resources, such as SSH security group rules
```

Comment:
Yes. I will spell it out a bit more in this comment, and then we can incorporate after discussion.

Right now, we are using CAPI to destroy the bootstrap machine. That is good because we can use the same code to delete the machine across all platforms.

The problems with this are:
* we need to implement further logic to clean up additional resources
* we need to keep the CAPI system running between provisioning resources and bootstrap destroy, which leads to problems with leaked processes (working on a fix)
* we will need to implement the ability to restart the capi system to run the standalone `openshift-install destroy bootstrap` command

For these reasons, it may be better to implement specific bootstrap destroy code for each platform.

Reply:
I would rather not. Unlike cluster destroy which can take a robust, blunderbuss approach to deleting things, deleting the bootstrap needs precision that will inevitably drift over time as we make changes to machine creation and forget this use case.

Reply:
After some further discussion, it seems like we just need to flesh out the CAPI approach a bit more:

>    we need to implement further logic to clean up additional resources

We need to do an accounting of what resources need to be destroyed per platform. We will probably need to extend the cloud platform provider interface to allow some resource specific logic. For example, we need to complete the design for how the SSH security rule for the bootstrap node would get destroyed. And to have a story for all resources we identify. (we can easily identify resources looking at terraform logs)

>    we need to keep the CAPI system running between provisioning resources and bootstrap destroy, which leads to problems with leaked processes (working on a fix)

As I mentioned, we do have a fix in mind for this (passing a top-level context down to the controllers)

>    we will need to implement the ability to restart the capi system to run the standalone openshift-install destroy bootstrap command

We need a general plan for how to handle this (have been discussing with @vincepri) but I think the implementation could potentially wait until closer to techpreview, or potentially even after.